Fix typing errors in torch.distributed.*, close issue #42967.

xuzhao9 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47534 Fix typing errors in torch.distributed.*, close issue #42967.**
* #47533 Fix typing errors in torch.distributed.nn.* directory.
* #47532 Fix typing errors in torch.distributed.distributed_c10d.*
* #47531 Fix type annotation errors in torch.distributed.* directory

Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

xuzhao9 committed:
Fix typing errors in torch.distributed.*, close issue #42967.

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 commented on torch/distributed/_pipeline/sync/checkpoint.py:
```diff
@@ -45,7 +45,7 @@
 # Types for shared memory between Checkpoint and Recompute.
 Recomputed = Tuple[TensorOrTensors, Tensors]  # (output, input_leaf)
-RNGStates = Tuple[ByteTensor, Optional[ByteTensor]]  # (cpu_rng_state, gpu_rng_state)
+RNGStates = Tuple[Tensor, Optional[Tensor]]  # (cpu_rng_state, gpu_rng_state)
```
Both torch.get_rng_state() and torch.cuda.get_rng_state(device) returns Tensor, not ByteTensor (in function save_rng_states()). So RNGStates should accept Tensor, not ByteTensor.

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

walterddr commented on torch/distributed/_pipeline/sync/stream.py:
```diff
@@ -104,7 +104,8 @@ def record_stream(tensor: torch.Tensor, stream: AbstractStream) -> None:
 
-        tensor.record_stream(as_cuda(stream))
+        # Typechecking: torch.cuda.Stream is incompatible with torch._C.Stream
+        tensor.record_stream(as_cuda(stream))  # type: ignore[arg-type]
```
should we change `tensor.record_stream` to return `AbstractStream` instead of `Stream` then?

xuzhao9 replied:
tensor.record_stream returns `None` and takes an argument of `torch._C.Stream`. This function is defined in torch._C.__init__.pyi, where we don't have the definition of `AbstractStream`. Neither can I find its related binding in C++. So I will leave it as-is until the next clean-up.

walterddr commented on torch/distributed/rendezvous.py:
```diff
@@ -1,12 +1,15 @@
     from urllib.parse import urlparse, urlunparse
 except ImportError:
-    from urlparse import urlparse, urlunparse
+    # Disable typechecking for Python2 modules
+    from urlparse import urlparse, urlunparse  # type: ignore
```
should we remove it since we no longer support python2
```suggestion
    raise ImportError("urllib cannot be found, urlparse from python2 is no longer supported")
```

xuzhao9 replied:
Good point. Fixed as suggested.

walterddr reviewed (approved):
looks good. minor 2 comments please feel free to address or ignore till the next clean up.

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."




[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."


Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."


Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."


Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."


Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

[ghstack-poisoned]

xuzhao9 committed:
Update on "Fix typing errors in torch.distributed.*, close issue #42967."


Differential Revision: [D24952497](https://our.internmc.facebook.com/intern/diff/D24952497)

[ghstack-poisoned]
